ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp

```diff
@@ -284,10 +296,56 @@ static int HandleVnodeOperation(
                 }
             }
         }
+        else if (ActionBitIsSet(action, KAUTH_VNODE_DELETE))
+        {
+            if (!TrySendRequestAndWaitForResponse(
+                    root,
+                    MessageType_KtoU_NotifyDirectoryPreDelete,
+                    currentVnode,
+                    pid,
+                    procname,
+                    &kauthResult,
+                    kauthError))
+            {
+                goto CleanupAndReturn;
+            }
+        }
     }
     else
     {
-        if (ActionBitIsSet(
+        if (ActionBitIsSet(action, KAUTH_VNODE_DELETE))
+        {
+            if (!TrySendRequestAndWaitForResponse(
+                    root,
+                    MessageType_KtoU_NotifyFilePreDelete,
+                    currentVnode,
+                    pid,
+                    procname,
+                    &kauthResult,
+                    kauthError))
+            {
+                goto CleanupAndReturn;
+            }
+            
+            // This delete could be part of a rename, and so we must hydrate the file now to
+            // ensure it's hydrated post-rename
+            if (kauthResult == KAUTH_RESULT_DEFER &&
```

Comment:
How could the result currently be anything but defer?

Reply:
With VFSForGit it will always be defer, but if a provider decided to block the delete of an empty placeholder it could be DENY.

(The idea here was to only hydrate if the delete is going to be allowed, let me know what you think)

Reply:
See related comment, I'll refactor this code to avoid duplication (and to hydrate first)